walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,134 +11,200 @@
-        intersection(availableTimesOptionals, availableTimes, duration);
+    // finds time ranges that does not work for at least one required attendee
+    for (Event event : events) {
+      TimeRange eventTimeRange = event.getWhen();
 
-    // Return time if any which works for both required and optional attendees
-    if (!availableTimesBothAttendees.isEmpty()) {
-      return availableTimesBothAttendees;
+      Collection<String> eventAttendees = event.getAttendees();
+      if (!Collections.disjoint(attendees, eventAttendees)) {
+        busyTimes.add(eventTimeRange);
+      }
     }
 
-    return availableTimes;
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    freeTimes = findFreeTimesFromBusy(busyTimes, duration);
+
+    return freeTimes;
   }
 
-  /** Finds avaliable times for the given atteendee list, input collection and event duration */
-  private Collection<TimeRange> getAvailableTimes(
-      Collection<String> attendees, Collection<Event> events, int duration) {
-    // TODO: Make this method more optimal
-    Collection<TimeRange> availableTimes = new ArrayList<>();
+  /** Calculates for each minute in a day how many optional attendee are free */
+  private List<Integer> getAvailableTimesForOptionals(
+      Collection<String> optionalAttendees, Collection<Event> events, int duration) {
+    Map<TimeRange, Set<String>> busyTimesForOptionals = new HashMap<>();
 
-    // possible start and end times of the event
-    int start = 0;
-    int end = start + duration;
-
-    // start and end time of the free period when event can occur
-    int availablePeriodStart = -1;
-    int availablePeriodEnd = -1;
-
-    // Consider every possible time period and see if it works for attendees
-    while (end <= TimeRange.END_OF_DAY) {
-      TimeRange currentTimeRange = TimeRange.fromStartDuration(start, duration);
-      boolean works = true;
-
-      // Comparing possible time range to the events to see if there is any overlaps
-      for (Event event : events) {
-        TimeRange eventTimeRange = event.getWhen();
-        if (currentTimeRange.overlaps(eventTimeRange)) {
-          Collection<String> eventAttendees = event.getAttendees();
-          if (!Collections.disjoint(attendees, eventAttendees)) {
-            works = false;
-          }
+    // for each event, finds which optional attendees are busy
+    for (Event event : events) {
+      TimeRange eventTimeRange = event.getWhen();
+
+      Collection<String> eventAttendees = event.getAttendees();
+      Set<String> busyOptionals = new HashSet<>(optionalAttendees);
+      busyOptionals.retainAll(eventAttendees);
+
+      if (busyOptionals.size() > 0) {
+        if (busyTimesForOptionals.containsKey(eventTimeRange)) {
+          busyTimesForOptionals.get(eventTimeRange).addAll(busyOptionals);
+        } else {
+          busyTimesForOptionals.put(eventTimeRange, busyOptionals);
         }
+      }
+    }
+
+    List<Set<String>> setsOfIptionals = new ArrayList<>();
```

Comment:
Nit: spelling, `setsOfOptionals`. Also, what this variable represents isn't very clear.